Client/Frontend/Browser/Search/BraveSearchManager.swift

```diff
@@ -0,0 +1,293 @@
+        guard let components = URLComponents(url: url, resolvingAgainstBaseURL: false),
+              let queryItem = components.valueForQuery("q") else { return nil }
+        
+        self.url = url
+        self.query = queryItem
+        self.domainCookies = cookies.filter { $0.domain == url.host }
+        if domainCookies.first(where: { $0.name == "fallback" })?.value != "1" {
+            return nil
+        }
+    }
+    
+    /// A call is made to the Brave Search api with query we performed.
+    /// It returns  details that are used in call to the fallback search engine.
+    func shouldUseFallback(completion: @escaping (BackupQuery?) -> Void) {
+        guard var canAnswerURLComponents =
+                URLComponents(string: url.domainURL.absoluteString) else {
+            completion(nil)
+            return
+        }
+        
+        canAnswerURLComponents.scheme = "https"
+        canAnswerURLComponents.path = "/api/can_answer"
+        canAnswerURLComponents.queryItems = [.init(name: "q", value: query)]
+        
+        guard let url = canAnswerURLComponents.url else {
+            completion(nil)
+            return
+        }
+        
+        var request = URLRequest(url: url,
+                                 cachePolicy: .reloadIgnoringLocalAndRemoteCacheData,
+                                 timeoutInterval: 5)
+
+        let cookieStorage = HTTPCookieStorage()
+        domainCookies.forEach { cookieStorage.setCookie($0) }
+        
+        let domainCookies = domainCookies.filter { $0.domain == canAnswerURLComponents.host }
+        let headers = HTTPCookie.requestHeaderFields(with: domainCookies)
+        headers.forEach {
+            request.setValue($0.value, forHTTPHeaderField: $0.key)
+        }
+        
+        // Important, URLSessionDelegate must have been implemented here
+        // to handle request authentication.
+        URLSession(configuration: .ephemeral, delegate: self, delegateQueue: .main)
+            .dataTaskPublisher(for: request)
+            .tryMap { output -> BackupQuery in
+                guard let response = output.response as? HTTPURLResponse,
+                      response.statusCode >= 200 && response.statusCode < 300 else {
+                    throw "Invalid response"
+                }
+     
+                let canAnswerResponse = try JSONDecoder().decode(BackupQuery.self, from: output.data)
+                
+                return canAnswerResponse
+            }
+            .sink(receiveCompletion: { status in
+                switch status {
+                case .failure(let error):
+                    log.error("shouldUseFallback error: \(error)")
+                    completion(nil)
+                case .finished:
+                    // Completion is called on `receiveValue`.
+                    break
+                }
+            }, receiveValue: { canAnswer in
+                completion(canAnswer)
+            })
+            .store(in: &cancellables)
```

Comment:
We need to call `session?.finishTasksAndInvalidate()` after `.store(...)` to prevent a leak. https://developer.apple.com/documentation/foundation/urlsession/1407428-finishtasksandinvalidate